### PR TITLE
feat: add common observable interface for hashers (APP-100)

### PR DESCRIPTION
**Pull request by anatolykislov**
- Added common observable interface CommitObservable for boosting of a performance. All hashers now receive per commit info synchronously with pre calculated diff. In such way we can traverse commit graph and calculate diff only once per commit. 
- Moved facts logic to separate hasher.
- Updated tests.
- CodeLongevity doesn't prepared for this interface yet. (How can we do it?)

Codes of facts will be updated in other PR.

### CODE REVIEW

File: src/main/kotlin/app/hashers/CommitHasher.kt [hunk 1]
```diff
@@ -3,240 +3,99 @@
...
+        val throwables = mutableListOf<Throwable>()
 
-        // Hash added and missing server commits and send them to server.
-        hashAndSendCommits()
+        observable
+            .takeWhile { new ->  // Hash until last known commit.
+                new.rehash != lastKnownCommit?.rehash
+            }
+            .filter { commit ->  // Don't hash known.
+                knownCommits.isEmpty() || !knownCommits.contains(commit)
```
Comment (asurkov) [inline src/main/kotlin/app/hashers/CommitHasher.kt]: how could we reach it if we do takeWhile() until the lastKnownCommit?

File: src/main/kotlin/app/hashers/CommitObservable.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import org.eclipse.jgit.diff.RawText
+import org.eclipse.jgit.errors.MissingObjectException
+import org.eclipse.jgit.lib.ObjectId
+import org.eclipse.jgit.revwalk.RevCommit
+import org.eclipse.jgit.revwalk.RevWalk
+import org.eclipse.jgit.util.io.DisabledOutputStream
+import java.nio.charset.Charset
+
+object CommitObservable {
+    fun getObservable(git: Git, repo: Repo) = Observable
```
Comment (asurkov) [inline src/main/kotlin/app/hashers/CommitObservable.kt]: It's a tautology to have CommitObservable.getObservable(). You may want to rename it to something like Commits.asObservable().

File: src/main/kotlin/app/hashers/CommitObservable.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                }
+                // Commits are combined in pairs, an empty commit concatenated
+                // to calculate the diff of the initial commit.
+                subscriber.onNext(Commit())
+            } catch (e: Exception) {
+                Logger.error("Commit producing error", e)
+                subscriber.onError(e)
+            }
+            subscriber.onComplete()
+        }  // TODO(anatoly): Rewrite diff calculation in non-weird way.
```
Comment (asurkov) [inline src/main/kotlin/app/hashers/CommitObservable.kt]: It's a funny TODO item but not clear what exactly you want to do here. It's up to you to replace on something more descriptive or just ignore it.

File: src/main/kotlin/app/hashers/RepoHasher.kt [hunk 4]
```diff
@@ -38,8 +47,16 @@ class RepoHasher(private val localRepo: LocalRepo, private val api: Api,
             getRepoFromServer()
 
             // Hash by all plugins.
-            CommitHasher(localRepo, repo, api, git).update()
-            CodeLongevity(localRepo, repo, api, git).update()
+            val observable = CommitObservable.getObservable(git, serverRepo)
+                                             .publish()
+            CommitHasher(localRepo, serverRepo, api, rehashes)
+                .updateFromObservable(observable)
```
Comment (asurkov) [inline src/main/kotlin/app/hashers/RepoHasher.kt]: I would probably rename the method to chainWith(observable) or JSish then(observable) or you could wrap that by some fancy interfaces like (names and code are sketchy)
interface Thenable {
  then(commits: Observable);
};
FunHasher : Thenable {
  override then(Observable)
}
and then
Commits.asObservalbe().
then(FunHasher).
then(CommitHasher)

File: src/main/kotlin/app/hashers/CommitObservable.kt [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                }
+        }
+    }
+
+    private fun getContentByObjectId(git: Git,
+                                     objectId: ObjectId): List<String> {
+        return try {
+            git.repository
+                .open(objectId).bytes.toString(Charset.defaultCharset())
+                .split('\n')
```
[the remark this answers is not in the document]
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CommitObservable.kt]: Ok.

File: src/main/kotlin/app/hashers/CommitHasher.kt [hunk 6]
```diff
@@ -3,240 +3,99 @@
...
-                            facts.add(Fact(repo, FactKey.COMMITS_DAY_WEEK +
-                                day, count.toDouble(), author))
-                        }
-                    }
-                }
-                postFactsToServer(facts)
-            })
+    init {
+        // Delete locally missing commits from server. If found at least one
+        // common commit then next commits are not deleted because hash of a
```
[the remark this answers is not in the document]
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CommitHasher.kt]: Yes.

File: src/main/kotlin/app/hashers/CommitHasher.kt [hunk 7]
```diff
@@ -3,240 +3,99 @@
...
+        // commit calculated including hashes of its parents.
+        val firstOverlapCommitRehash = findFirstOverlappingCommitRehash()
+        val deletedCommits = serverRepo.commits
+            .takeWhile { it.rehash != firstOverlapCommitRehash }
+        deleteCommitsOnServer(deletedCommits)
```
[the remark this answers is not in the document]
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CommitHasher.kt]: Let's stay on current version of a protocol.

File: src/main/kotlin/app/hashers/CommitHasher.kt [hunk 8]
```diff
@@ -3,240 +3,99 @@
...
-                             old = DiffContent(old, edits.map { edit ->
-                                 DiffRange(edit.beginA, edit.endA) }),
-                             new = DiffContent(new, edits.map { edit ->
-                                 DiffRange(edit.beginB, edit.endB) }))
-                }
+        val serverHistoryRehashes = serverRepo.commits
+                                              .map { commit -> commit.rehash }
+                                              .toHashSet()
+        return rehashes.firstOrNull { rehash ->
+            serverHistoryRehashes.contains(rehash)
```
[the remark this answers is not in the document]
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CommitHasher.kt]: I'm looking for first mapping of M(local) to N(server) commits not M(local) to 1. HashSet speed up the process.

File: src/main/kotlin/app/hashers/CommitHasher.kt (hunk 1 above)
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CommitHasher.kt]: Server could miss some of the commits.

File: src/main/kotlin/app/hashers/CommitObservable.kt (hunk 2 above)
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CommitObservable.kt]: Changed name to CommitCrawler, method name the same.

File: src/main/kotlin/app/hashers/CommitHasher.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (asurkov) [reply src/main/kotlin/app/hashers/CommitHasher.kt]: Not sure I follow it, anyway, this code wasn't introduced by this pull request, so I guess I can revisit this part later and file followup issue if needed.

**Pull request merged**